### ISSUE TITLE
Crashlytics lane fails on first run

### ISSUE THREAD

**Issue by vincentbrison** [fastlane/fastlane#12566]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When using [crashlytics action](https://docs.fastlane.tools/actions/crashlytics/) to upload an Android build to crashlytics beta, the first run always crash.

Given the stacktrace, it looks like a temp file does not exist when `com.crashlytics.tools` jar is running, causing the crash.

We use the following configuration:
```
lane :beta do
    gradle(task: 'assemble', build_type: 'beta')
    crashlytics(
        api_token: 'an_api_token',
        build_secret: 'a_build_secret',
        groups: 'a_group_of_testers',
        notes: 'some changelog'
    )
end
```

The issue only occurs when at the first time we run `fastlane beta`, when `notes` parameter is set. On the second run, the temp file is here, and the task complete successfully.

#### Our understanding
From the sources of [crashlytics.rb](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/crashlytics.rb), when using `notes` parameter, a temp file is used to store the notes, before passing the path of this file to crashlytics dev tools which itself do the upload to beta store.
From [crashlytics_helper.rb](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/helper/crashlytics_helper.rb), this temp file is generated using [ruby `tempfile` class](https://ruby-doc.org/stdlib-1.9.3/libdoc/tempfile/rdoc/Tempfile.html).

So it may looks like the first time we use `fastlane beta`, something is cleaning the temp folder before the upload by crashlytics dev tools. The only difference between the first run of `fastlane beta` and subsequent runs is that [crashlytics_helper.rb](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/helper/crashlytics_helper.rb) will call `download_android_tools` to download crashlytics dev tools jar.
We forked fastlane to add logs over the temp folder used by uby `tempfile` class (usually /tmp folder), to see if it may be related to this download. We found that https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/helper/crashlytics_helper.rb#L88 was cleaning the temp folder.

The following issue https://github.com/fastlane/fastlane/issues/6325 is probably related to this issue. Systems without state like docker .img will systematically meet this issue since they will always be in the state of the first run. It may explain why people are randomly meet this error. Finally there is nothing random here, it is just related to the presence of not of crashlytics dev tools jar.

#### Recreate the issue
You can recreate easily the issue, by calling the following command to simulate a first run:
`rm -rf ~/Library/CrashlyticsAndroid && bundle exec fastlane beta`, using a dummy Android project to upload to crashlytics beta.

#### Possible fixs
A quick fix would be to download the crashlytics dev tools before generating the notes files into the tmp folder, but I am not a ruby nor a fastlane expert to say so.
To quick fix the issue on your side, we directly versioned crashlytics dev tools with our project, and passed it to fastlane with `crashlytics_path` parameter.

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>[10:09:28]: fastlane finished with errors

[!] ERROR - Crashlytics Developer Tools error.
java.io.FileNotFoundException: /var/folders/w9/dklp54mj4ns1cmk5_9zqf54h0000gn/T/changelog20180521-59451-1c4qa7c (No such file or directory)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:126)
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:116)
        at com.google.common.io.ByteSource$AsCharSource.openStream(ByteSource.java:435)
        at com.google.common.io.CharSource.read(CharSource.java:162)
        at com.google.common.io.Files.toString(Files.java:367)
        at com.crashlytics.tools.android.DeveloperTools.getPropertyValueOrValueFromPropertyPath(DeveloperTools.java:743)
        at com.crashlytics.tools.android.DeveloperTools.processProperties(DeveloperTools.java:662)
        at com.crashlytics.tools.android.DeveloperTools.processArgsInternal(DeveloperTools.java:348)
        at com.crashlytics.tools.android.DeveloperTools.main(DeveloperTools.java:273)
Exception in thread "main" com.crashlytics.tools.android.exception.PluginException: Crashlytics Developer Tools error.
        at com.crashlytics.tools.android.DeveloperTools.processArgsInternal(DeveloperTools.java:356)
        at com.crashlytics.tools.android.DeveloperTools.main(DeveloperTools.java:273)
Caused by: java.io.FileNotFoundException: /var/folders/w9/dklp54mj4ns1cmk5_9zqf54h0000gn/T/changelog20180521-59451-1c4qa7c (No such file or directory)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:126)
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:116)
        at com.google.common.io.ByteSource$AsCharSource.openStream(ByteSource.java:435)
        at com.google.common.io.CharSource.read(CharSource.java:162)
        at com.google.common.io.Files.toString(Files.java:367)
        at com.crashlytics.tools.android.DeveloperTools.getPropertyValueOrValueFromPropertyPath(DeveloperTools.java:743)
        at com.crashlytics.tools.android.DeveloperTools.processProperties(DeveloperTools.java:662)
        at com.crashlytics.tools.android.DeveloperTools.processArgsInternal(DeveloperTools.java:348)
        ... 1 more

[10:09:28]: WARNING: fastlane requires your locale to be set to UTF-8. To learn more go to https://docs.fastlane.tools/getting-started/ios/setup/#set-up-environment-variables</pre>
</details>

### Environment

<details>
  <pre><details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                |
| --------------------------- | ------------------------------------ |
| OS                          | 10.13.4                              |
| Ruby                        | 2.2.4                                |
| Bundler?                    | false                                |
| Git                         | git version 2.17.0                   |
| Installation Source         | ~/.fastlane/bin/bundle/bin/fastlane  |
| Host                        | Mac OS X 10.13.4 (17E202)            |
| Ruby Lib Dir                | ~/.fastlane/bin/bundle/lib           |
| OpenSSL Version             | OpenSSL 1.0.2g  1 Mar 2016           |
| Is contained                | false                                |
| Is homebrew                 | true                                 |
| Is installed via Fabric.app | false                                |
| Xcode Path                  | /Library/Developer/CommandLineTools/ |
| Xcode Version               |                                      |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# This file contains the fastlane.tools configuration
# You can find the documentation at https://docs.fastlane.tools
#
# For a list of all available actions, check out
#
#     https://docs.fastlane.tools/actions
#
# For a list of all available plugins, check out
#
#     https://docs.fastlane.tools/plugins/available-plugins
#

# Uncomment the line if you want fastlane to automatically update itself
# update_fastlane

default_platform(:android)

platform :android do
  desc "Runs all the tests"
  lane :test do
    gradle(task: "test")
  end

  lane :buildDebug do
    gradle(task: 'assemble', build_type: 'Debug')
  end

  lane :buildBeta do
      gradle(task: 'assemble', build_type: 'Beta')
   end

  lane :buildRelease do
    gradle(task: 'assemble', build_type: 'Release')
  end

  lane :betaChangelog do
    changelog = "Version : " + gradle(task: 'printVersionName', flags: '-q')
    # Generate the changelog based on commit messages since your last tag
    changelog = changelog + changelog_from_git_commits
  end

  desc "Submit a new Beta Build to Crashlytics Beta"
    lane :beta do
        gradle(task: 'assemble', build_type: 'boPreProdBeta')
        crashlytics(
            api_token: ' 3c25f278836e1f6e5b8469cd982d3ca4711458a2',
            build_secret: '5d1f1e46d6188491f64f00891b8f3bf43e1f730d0c4bb9f916e5a9c1a20900fc',
            groups: 'keyopstech-internal',
            notes: betaChangelog,
            crashlytics_path: "./crashlytics/crashlytics-devtools.jar"
        )
    end

  desc "Submit a new Release Build to Crashlytics Beta"
    lane :betaRelease do
      gradle(task: 'assemble', build_type: 'boProdRelease')
      crashlytics(
          api_token: ' 3c25f278836e1f6e5b8469cd982d3ca4711458a2',
          build_secret: '5d1f1e46d6188491f64f00891b8f3bf43e1f730d0c4bb9f916e5a9c1a20900fc',
          groups: 'keyopstech-internal',
          notes: betaChangelog,
          crashlytics_path: "./crashlytics/crashlytics-devtools.jar"
        )
    end
end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
json_key_file("") # Path to the json secret file - Follow https://docs.fastlane.tools/actions/supply/#setup to get one
package_name("com.zol.keyops") # e.g. com.krausefx.app

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.95.0  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.2        |
| claide                    | 1.0.2        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.2.5        |
| xcodeproj                 | 1.5.9        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.6        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| public_suffix             | 2.0.5        |
| tty-screen                | 0.6.4        |
| tty-cursor                | 0.5.0        |
| tty-spinner               | 0.8.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| commander-fastlane        | 4.4.6        |
| faraday                   | 0.12.1       |
| unf_ext                   | 0.0.7.4      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20170404 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| gh_inspector              | 1.1.3        |
| json                      | 1.8.1        |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.2.1        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 0.0.4        |
| bundler                   | 1.14.6       |
| faraday_middleware        | 0.11.0.1     |
| naturally                 | 2.1.0        |
| simctl                    | 1.6.4        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| little-plugger            | 1.1.4        |
| logging                   | 2.2.2        |
| jwt                       | 1.5.6        |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.13.6       |
| emoji_regex               | 0.1.1        |
| highline                  | 1.7.10       |
| excon                     | 0.62.0       |
| CFPropertyList            | 3.0.0        |
| libxml-ruby               | 3.0.0        |
| plist                     | 3.4.0        |
| fastimage                 | 2.1.3        |
</details>

*generated on:* **2018-05-21**
</details></pre>
</details>

**Comment by joshfriend** [comment]
please re-open

**Comment by vincentbrison** [comment]
This is still an issue, it needs to be re-open.

**Comment by chuganzy** [comment]
The PR has been just merged! Thanks @joshdholtz reviewing it.
You can try it by modifying `gem "fastlane"` in your gemfile with:
```ruby
gem 'fastlane', :gitbub => 'fastlane', :ref => '6cf7c2afcd5b6532257ff23b797bd098a337c67b'
```
and run `bundle install`.

**Comment by chuganzy** [comment]
Looks the patch was just released in 2.114.0.
We've been using it internally for awhile and do not see the issue anymore!

**Comment by janpio** [comment]
If anyone still has this problem, please respond to this issue. Until then, we will consider this fixed and close this issue for now 👍
